Enum support

maxenglander opened the pull request:
Status: approved.
Resolves: #450.
Provides: MySQL-style enum support e.g. `CREATE TABLE shirt (size ENUM('small', 'large'))`. 

---
 
#### TODO

 * [x] Write code
   * [x] Store value as ints
   * [x] Allow updating/inserting with strings or ints
   * [x] Allow comparison (e.g. in where clauses) with strings or ints
   * [x] Convert stored int back to string when rendering query results
   * [x] Verify properly creates table from querying a table with an enum type
   * [x] Verify properly group by enum columns
   * [x] Support enum arithmetic
   * [ ] ~~Maybe support insert/update/comparison with other types (boolean, decimal...?)~~
 * [x] Write tests
 * [x] Verify existing tests not broken
 * [x] Provide end-user documentation
 * [x] Document grammar changes
 * [x] Provide a change-log entry
 * [x] Verify spelling
 * [x] Verify formatting

<a name="questions"></a>
#### QUESTIONS

MySQL enums have (at least) a couple weird quirks, which appear to apply at least partially to Oracle, and not at all to Postgres. I'm hoping the H2 team can provide me with some guidance about how I should handle these cases. Make the MySQL logic the default behavior? Enable the MySQL logic when the MYSQL dialect is requested? Do not port the logic to H2?

##### MySQL enums are 1-indexed and allow empty strings

For a MySQL column of type `enum('a','b','c')`, the integer value of `'a'` is `1`, `'b'` `2`, and so on. A value of `0` or `''` may be set/inserted, and will not map to either `'a'`, `'b'`, `'c'`.

It appears [Oracle](https://docs.oracle.com/cd/E17952_01/mysql-5.0-en/enum.html) has similar behavior.

[Postgres](https://www.postgresql.org/docs/9.1/static/datatype-enum.html), as far as I can tell, will _not_ let you insert empty strings.

My two cents is that it would be beneficial to MySQL and Oracle users to implement this quirk in H2, since both MySQL and Oracle share the same syntax for the enum type generally speaking, and this behavior in particular. If/when support for the Postgres syntax is added, it won't be too difficult to disable this quirk for enum-typed columns defined via that syntax.

##### MySQL enum arithmetic is weird

With MySQL enums, you can do `update a_table set an_enum_column = an_enum_column + 1`. This isn't weird, but what is a bit weird is that if a row  with `an_enum_column` is at the highest enum value of the column, adding `+ 1` will cause its value to loop back around to zero which, in MySQL's case, is an empty string (see above).

    mysql> CREATE TABLE card (rank INT, suit ENUM('clubs','hearts','spades','diamonds'));
    mysql> INSERT INTO card VALUES(0, 'diamonds');
    mysql> SELECT suit FROM card;
    +----------+
    | suit     |
    +----------+
    | diamonds |
    +----------+
    mysql> UPDATE card SET suit = suit + 1;
    mysql> SELECT suit FROM card;
    +------+
    | suit |
    +------+
    |      |
    +------+

Perhaps this by itself isn't so weird, but what _is_ weird is the additional property that if you do `update a_table set an_enum_column = an_enum_column - 1`, and a row has `an_enum_column` at zero (that is, an empty string), it will not loop back around the other way.

    mysql> SELECT suit FROM card;
    +------+
    | suit |
    +------+
    |      |
    +------+
    mysql> UPDATE card SET suit = suit - 1;
    mysql> SELECT suit FROM card;
    +------+
    | suit |
    +------+
    |      |
    +------+
    
I don't know whether or not [Oracle](https://docs.oracle.com/cd/E17952_01/mysql-5.0-en/enum.html)  has similar behavior.

[Postgres](https://www.postgresql.org/docs/9.1/static/datatype-enum.html) supports comparisons between members of a particular enum, but does not appear to support other types of arithmetic, or comparison to other types such as ints.

This quirk is documented in neither MySQL nor Oracle, as far as I can tell, so I feel less confident that it should be ported to H2.

maxenglander committed:
Enum type does not support precision or scale

maxenglander committed:
restrict permitted values for enum columns

maxenglander committed:
Merge branch 'master' into enum-support

maxenglander committed:
Add prepared statement tests for enum data type

grandinj commented:
that's a pretty good start. Some comments:

 - we need to use a List<>, not a Set<> to store the values, because we need their index, because...

 - having ValueEnum extend ValueString is not ideal. That means we store the values as actual strings.
 We should be storing them as the corresponding integer, determined by their position in the list.
 
 - and we need to sort them by ORDER BY clauses using their position in the list

maxenglander commented:
Thank you @grandinj! I'll incorporate your requested changes.

lukaseder commented:
Very interesting feature addition. Has anyone expressed interest in supporting PostgreSQL's enum type as well?

grandinj commented:
Postgrel's ENUM type should be a fairly small addition on top of this, once it's done. They are not all that different.

lukaseder commented:
@grandinj Indeed, that's how I see it. Perhaps this implementation could be implemented as a special case of PostgreSQL's i.e. there would be a system-generated (invisible?) enum type...

grandinj commented:
@lukaseder in theory, yes, but I prefer not too discourage contributors with extra requirements unnecessarily :-)

 I'm fairly sure that transformation could be bolted on later without much effort.

lukaseder commented:
Fair point :)

maxenglander committed:
store enums as ints

grandinj commented on h2/src/main/org/h2/value/Value.java:
```diff
@@ -167,6 +168,11 @@
+    /**
      * The number of value types.
      */
     public static final int TYPE_COUNT = TIMESTAMP_TZ;
```
TYPE_COUNT needs to be updated

grandinj commented on h2/src/main/org/h2/value/ValueEnum.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Implementation of the ENUM data type.
+ */
+public class ValueEnum extends ValueString {
```
because you are still extending ValueString, I'm pretty sure the storage on disk is going to store string values instead of integers

maxenglander replied:
Gotcha! Thanks for the tip. Will make sure to address.

grandinj reviewed (changes requested):
NIce progress!

maxenglander committed:
enum-support: drop ValueEnum.java (for now); fix originalSQL when parsing enum columns

maxenglander committed:
display enum labels while maintaining sort-on-int-value

maxenglander committed:
enum-support: handle enum comparisons with ints and strings

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4058,6 +4058,7 @@ private Column parseColumnWithType(String columnName) {
         }
         long precision = -1;
         int displaySize = -1;
+        java.util.Set<String> permittedValues = new HashSet<>();
```
more efficient to initialise this on-demand, i.e. leave it null to start with

maxenglander committed:
enum-support: fix bug where enumerators incorrectly assigned to column

maxenglander committed:
enum-support: clean up

maxenglander committed:
enum-support: merge master

maxenglander committed:
enum-support: fix NPE with Column copy

maxenglander committed:
enum-support: fix/add tests

maxenglander committed:
enum-support: fix Transfer.{read,write}Value for enum

maxenglander committed:
enum-support: create table (with enum column) from select statement

maxenglander committed:
enum-support: add support for math operations

maxenglander committed:
enum-support: merge master

maxenglander committed:
enum-support: fix bad exception throw

maxenglander committed:
enum-support: add more test cases

maxenglander commented:
Hi @grandinj,

I believe I've addressed your requested changes, and have covered the core enum behaviors and edge cases. I hope you'll take another look, and guide me on some [questions](#user-content-questions) I have related to MySQL enum behaviors that you may wish to (not) incorporate, or to limit within the MySQL dialect mode.

Thanks!
Max

grandinj commented on h2/src/main/org/h2/value/Transfer.java:
```diff
@@ -494,6 +494,20 @@ public void writeValue(Value v) throws IOException {
+            writeInt(ve.getInt());
+
+            break;
+        }
```
this code is used when communicating between client and server. In which case there is no need to send all of the enumerators. At most you need to send one integer value and one string value of the enumerator.

maxenglander replied:
OK @grandinj makes sense. However, in this case I think I will also have to pass along some identifier - e.g. table and column ids - so that `readValue` can find the complete set of enumerators via a `DataHandler` (or similar) interface? Let me know if that's not right.

grandinj replied:
Yes, that would work fine.

The other option would be to have two ValueEnum classes - one for the server side, and one for the client side. The client side one would only need to know the integer id and the string value, because it would only be used to pass data through the JDBC interface. The server side one would need the full set of enumerators,

grandinj commented on h2/src/main/org/h2/value/Transfer.java:
```diff
@@ -594,6 +608,14 @@ public Value readValue() throws IOException {
+                enumerators[i] = readString();
+            }
+            return ValueEnum.get(enumerators, readInt());
+        }
```
which will require some ickiness here, in that you'll need to construct a ValueEnum that does not know what it's complete set of enumerators are.
We face a similar problem with ValueLob and we get around it by passing in a DataHandler interface (really bad name).
We could do something similar with ValueEnum, passing in an interface which allows it request the full set of enumerators for itself on-demand.

maxenglander committed:
enum-support: in Transfer, use base ValueEnumBase when reading

maxenglander commented on h2/src/main/org/h2/value/Transfer.java:
```diff
@@ -594,6 +599,11 @@ public Value readValue() throws IOException {
+        case Value.ENUM: {
+            final int ordinal = readInt();
+            final String label = readString();
+            return ValueEnumBase.get(label, ordinal);
```
Hey @grandinj I've put the solution you described here in place - different `ValueEnum` for the client and server.

grandinj commented:
This is awesome work!

Please can we have a license statemtn send to the mailing list for our records, something like this:

http://h2database.com/html/build.html#providing_patches

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4127,6 +4128,8 @@ private Column parseColumnWithType(String columnName) {
         }
         long precision = -1;
         int displaySize = -1;
+        java.util.List<String> enumeratorList = null;
```
move this declaration inside the "if readIf("("" block

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4200,6 +4203,29 @@ private Column parseColumnWithType(String columnName) {
+                enumeratorList = new ArrayList<String>();
+                original += '(';
+                String enumerator0 = readString();
+                enumeratorList.add(enumerator0.toLowerCase().trim());
```
here and just below, rather call toUpperCase(Locale.ENGLISH), since we tend to make things uppercase by default, and to avoid locale issues with case shifting

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4220,8 +4246,10 @@ private Column parseColumnWithType(String columnName) {
+
         Column column = new Column(columnName, type, precision, scale,
-                displaySize);
+            displaySize, enumerators == null ? null : enumerators);
```
this can just be "enumerators" :-)

grandinj commented on h2/src/main/org/h2/table/Column.java:
```diff
@@ -133,8 +146,12 @@ public int hashCode() {
     }
 
+    public boolean isEnumerated() {
+        return enumerators != null && enumerators.length > 0;
```
no need to check the length - if we have a zero length array, we have other problems

maxenglander replied:
Hey @grandinj as I have things set up now `enumerators` will always be `NULL` for all non-`ENUM` data types. Maybe it would be better to simply check if `dataType.enumerated == true`? I'm going to make that change now.

With regards to your comment "if we have zero length array, we have other problems" - would you prefer I set things up so that `enumerators` is never `NULL`, even for non-`ENUM` data types?

grandinj replied:
checking for dataType.enumerated is fine. 

I think enumerators should be null for non-ENUM data types.  I'd rather have a NullPointerException immediately if code does something wrong, than have weird behaviour further down the code.

grandinj commented on h2/src/main/org/h2/value/DataType.java:
```diff
@@ -141,6 +141,11 @@
     public boolean caseSensitive;
 
     /**
+     * If permitted values are supports.
```
"If enumerated values are supported"

inikolaev commented on h2/src/test/org/h2/test/testScript.sql:
```diff
@@ -10588,4 +10588,76 @@ create table z.z (id int);
+--- ENUM support
+----------------
+
+--- ENUM basic operations
```
Is it possible to add a test for a custom type? Something like:

```sql
create type card_suit as enum('hearts', 'clubs', 'spades');
create table card (rank int, suit card_suit);
and etc...
```
I don't really know if it is necessary, but this is the use case I'm personally interested in and would love this to be working :)

maxenglander replied:
@inikolaev I tried adding this test but, unfortunately, it did not work. I would be happy to address this in a follow-on PR.

maxenglander committed:
enum-support: use toUpperCase(ENGLISH) instead of toLowerCase()

maxenglander committed:
enum-support: add docs

maxenglander commented:
@grandinj Thank you for the recent code review! I believe I have addressed all of your comments, including sending a message to the user group.

I did want to ask again about the [questions](#user-content-questions) above. I do not at all mind leaving these questions unaddressed, but I just wanted to make sure they hadn't escaped your attention.

lukaseder commented:
This would be nice as an addition to the existing implementation: https://github.com/h2database/h2database/issues/654
